Attribute Data System (part I)

Z3rca opened the pull request:
Disclaimer, this is not the final system, rather a stop in between, to check if we agree on the common vision and test if there are already problems with it. 

### Description 
The AttributeDataSystem is the foundation for storing relevant Party and Character information of the Old-World mod during Campaign Gameplay and Combat. Custom Information is stored by this System into a dictionary, is read during combat, and assigns these data to the agent. e.g., that a character is being a mage, and it is provided a certain skill set. it also could contain faction or additional profession attributes. The AttributeDataSystem is accountable for Party data only but should contain these data exhaustive for all future modifications for parties of the Old-World Mod. 

It makes use of Taleworlds data saving and loading system as well as TaleWorlds CampaignEvents.

It involves the following classes:

- `PartyAttributeManager` : Manages attribute data on all agents and characters. In later stages, the Manager should be able to read out The old World XML data for characters or units, providing additional effects , spells other attributes we plan to include.
- `PartyAttribute`: stores information related to a party. Extendible and changeable in future 
- `StaticAttribute`: stores information of single characters and unit roster elements. Extendible and changeable in future.
- `StaticAttributeMissionLogic`: reads out `PartyAttribute`s of involved parties during Combat and assigns this information correctly to an agent with an added `StaticAgentComponent`. 
- `StaticAgentComponent`: contains a data of a single `StaticAttribute`. In Later stages, this Component should also enable and assign Abilities, and status effects on agents.

### functions

In the current state, it can create dummy data for Lord parties, Rogue parties and Regular parties like caravans or villagers during Game start. Destroyed and added parties were also updated in the dictionary. The information can be saved and load. It also increases a Winds of Magic Resource while moving around on Campaign Map. During Combat, the involved Party information are read out and assigned to agents. 

### Known Issues:

- Once changes were made on Party Attributes or StaticAttributes, the Savefile is broken, or values are irreparable. The system also requires a validity check for save files.
- Some parties or character changes or party changes are still not detected correctly, due to missing TaleWorld events. Therefore, some information passes are left out or look “hacky”.
- Rogue Parties only contain a single attribute for all troops even when we try to change or add additional rosters to them.




UPDATE: 

### first Functional read out of Status effects and Abilities:

in `PartyAttributeManager` and  `StaticAttribute` only now necessary items are stored. In addition Winds of Magic is now added as a resource, which is gained over campaign time, and is depleted by casting spells. Though theoretically possible, there is currently no feature adding fireball abilities to a character through learning. The Bright wizard wanderer, and the player choosing at the beginning of the campaign "BrightWizard Pupil" will get access to a fireball.  For testing purposes, before starting a new game, Rogue Parties can be given the "undead" attribute on launching a new campaign, which applies all rules of the undead Moral System. All Spells do cause in campaign battles the application of Status effects, as well as bandits, start to crumble, when the undead attribute is provided.

This test shows, the ease of implementing new attributes, Status effects, or Spells to that system. new Attributes can be added, removed, and read out at any time. In future, more meaningful resources like faith or other attributes can be added. While it makes use of `AgentExtensions`, the system is only coupled to our own classes, and not relying tremendously on Taleworlds Code.

known issues:
During Campaign Map a HUD is missing, which shows the current amount of Winds of Magic. Currently, only per day, the current available Winds of Magic is shown in form of a text message. 
Since Companions currently can't use Winds of Magic, I think that the system requires a little overwork then. 

Note:
TOW_CoreAssets have to be brought on the state of GUICampaignMagic. The Changes to the XML had to be made here. When this branch is merged, please also merge GUICampaignMagic

Z3rca committed:
added former files to the project

Z3rca committed:
enable folder in cs project file

Z3rca committed:
added event handling for fighting parties

unfortunately do not detect player party when getting into fights. works for other perfect.

important: changed the way the ids are stored, to make it easier readable.

Z3rca committed:
working on Player related Campaign Events

Z3rca committed:
WIP single Entity static Attributes

Finding all involved parties works, now we have to come up with solution for all parties and their given troops

Z3rca committed:
working in progress , create new static attribute data container

they cant be saved currently

Z3rca committed:
refined saving structure

Z3rca committed:
Changed name of Attribute

Z3rca committed:
magic user check

Z3rca committed:
removed function of Attributebehavior script

did cause problems on loading party attributes
forgot to push attribute for magic users

Z3rca committed:
work in progress: access relevant information during combat

Z3rca committed:
adjustment on campaignBehavior

Z3rca committed:
attempt of assigning attributes on  agents

Z3rca committed:
changing from defender and attackers to "active" parties

Z3rca committed:
removed logic from StatusEffect system

Z3rca committed:
forgotten submodule logic

Z3rca committed:
attributes work for players

Z3rca committed:
workaround for bandit parties

for some reasons they are neither recognized as heroes nor regular troops. I have to use a workaround here

Z3rca committed:
Merge branch 'CampaignAttributesNew' into CampaignAttributeRebase

Z3rca committed:
managed assign sucessfully rogue party members and players to system

Z3rca committed:
working on companion Attributes and regular troop attributes

Z3rca committed:
changed Party Types to Enums

also changed the way heroes are added to a party

Z3rca committed:
changed breaks to contiues

Z3rca committed:
added singleton pattern for AttributeSystemManager

Z3rca committed:
changed from AgentCreated to AgentBuild

this allows to access parties of an agent

Z3rca committed:
fixed some last bugs for merge

Z3rca committed:
tidied up AttributeSystemManager

Z3rca committed:
readded necessary assignment

Z3rca committed:
removed unnessary script

Z3rca committed:
removed unnessary code lines

Z3rca committed:
removed unused scripts

Z3rca committed:
more tidying

Z3rca committed:
renamed to PartyAttributeManager

Z3rca committed:
forgotten changes in other files, added description texts

Z3rca committed:
renamed folder in project

Z3rca committed:
more description and tidying

Z3rca committed:
changed namespace and unnessary usings

Yaroslav1996 commented on CSharpSourceCode/AttributeDataSystem/PartyAttributeManager.cs:
```diff
@@ -0,0 +1,439 @@
+        {
+            
+        }
+        
```
Empty method?

Z3rca replied:
will be removed. thanks

Yaroslav1996 commented on CSharpSourceCode/AttributeDataSystem/PartyAttributeManager.cs:
```diff
@@ -0,0 +1,439 @@
+            if (party.IsBandit)
+            {
+                StaticAttribute staticAttribute = new StaticAttribute();
+                staticAttribute.race = "bandit";
```
bandit race? Maybe rename race to tag or sth, I think race should be assigned to a certain character, not the entire party

Z3rca replied:
the static attribute to each unit which origins from a troop tree, or a hero or companion. in case of bandits currently, they only have a single static attribute, since I couldn't access its troop tree.

Yaroslav1996 commented on CSharpSourceCode/Battle/StatusEffects/StatusEffectMissionLogic.cs:
```diff
@@ -28,6 +32,93 @@ public override void OnAgentCreated(Agent agent)
+                TOWCommon.Say(defenderAttributes[0].WindsOfMagic.ToString());
+            }#1#
+        }*/
+
```
Remember to remove commented code before merging

Z3rca committed:
changed race and culture attribute

and filled it with dummy values
also removed unused fill timer class

Z3rca committed:
removed comment outcommented line

Yaroslav1996 commented on CSharpSourceCode/AttributeDataSystem/StaticAttribute.cs:
```diff
@@ -0,0 +1,41 @@
+        //dummy data types
+        [SaveableField(0)] public string race;  //mostly hardcoded currently needs to be assigned via xml or similar also counts for many other values in this class
+        [SaveableField(1)] public string culture; // eg. undead, Vampire, Human
+        [SaveableField(2)]public bool MagicUser;   
```
Wouldnt IsMagicUser be a better name for bool?

Z3rca replied:
agree

Z3rca committed:
changed Is for MagicUser bool

tristanka commented on CSharpSourceCode/AttributeDataSystem/PartyAttribute.cs:
```diff
@@ -0,0 +1,79 @@
+        public List<StaticAttribute> RegularTroopAttributes = new List<StaticAttribute>();
+
+        [SaveableField(8)] public PartyType PartyType;
+        [SaveableField(9)] public int numberOfRegularTroops;
```
Capital N `NumberOfRegularTroops` to maintain a consistent convention for these public fields?

Also, for lines 27, 28, and 29, is there a reason the `SaveableField` attribute is on the same line as the field? I personally think it's better to have it above the field but I'd like it to at least be consistent.

Z3rca replied:
Was removed in recent version

tristanka commented on CSharpSourceCode/AttributeDataSystem/PartyAttribute.cs:
```diff
@@ -0,0 +1,79 @@
+            base.DefineEnumTypes();
+            AddEnumDefinition(typeof(PartyType), 1);
+        }
+        
```
Now, I _know_ this is nitpicky, so bear with me, but I figure I might as well mention it anyway if you're in the code making changes. Can we use consistent block separations/new lines? i.e. remove the newline on 26, 41, 58, and 77?

tristanka commented on CSharpSourceCode/AttributeDataSystem/PartyAttribute.cs:
```diff
@@ -0,0 +1,79 @@
+        [SaveableField(1)]
+        public Hero Leader;
+        [SaveableField(2)]
+        public string id;           //partybase.
```
What is this `//partybase.` comment for? If it's meaningful, can you add more information? It's not clear to me what it's saying. If it's a "foreign key" to a Party, maybe rename the variable to something like `partyId` and remove the comment.

Z3rca replied:
done

tristanka commented on CSharpSourceCode/AttributeDataSystem/PartyAttributeManager.cs:
```diff
@@ -0,0 +1,437 @@
+    /// </summary>
+    public class PartyAttributeManager: CampaignBehaviorBase
+    {
+        private static  PartyAttributeManager  _instance = new PartyAttributeManager();
```
Two spaces between `static` and `PartyAttributeManager`, and two spaces between `PartyAttributeManager` and `_instance`. Change these two one space?

Z3rca replied:
rider sometimes do strange things... o.O

tristanka commented on CSharpSourceCode/AttributeDataSystem/PartyAttributeManager.cs:
```diff
@@ -0,0 +1,437 @@
+        private PartyAttributeManager()
+        {
+            
+        }
```
New line here?

Z3rca replied:
changed

tristanka commented on CSharpSourceCode/AttributeDataSystem/PartyAttributeManager.cs:
```diff
@@ -0,0 +1,437 @@
+        
+    }
+    
+        public class BattleAttributesArgs : EventArgs
```
This class is indented too much

tristanka commented on CSharpSourceCode/AttributeDataSystem/PartyAttributeManager.cs:
```diff
@@ -0,0 +1,437 @@
+        {
+            dataStore.SyncData("_partyAttributes", ref _partyAttributes);
+        }
+        
```
Delete these new lines? In general, just be consistent with the number of newlines between different structures. e.g. delete the new line on 37, 82, 83... add a new line after 92, etc

tristanka commented on CSharpSourceCode/AttributeDataSystem/StaticAttribute.cs:
```diff
@@ -0,0 +1,41 @@
+        [SaveableField(6)]public List<string> MagicEffects = new List<string>();
+        [SaveableField(7)]public string id;
+        [SaveableField(8)]public int number;
+        [SaveableField(9)]public PartyBase AssignedParty;
```
These `SaveableField` attributes are all before the fields, but `PartyAttribute`, they're on the next line. Can we do the next line style instead?

tristanka commented on CSharpSourceCode/AttributeDataSystem/StaticAttribute.cs:
```diff
@@ -0,0 +1,41 @@
+        [SaveableField(2)]public bool IsMagicUser;   
+        [SaveableField(3)] public int faith;    
+        [SaveableField(4)]public string status;
+        [SaveableField(5)]public List<string> SkillBuffs =new List<string>();
```
`=new` -> `= new`?

tristanka commented on CSharpSourceCode/AttributeDataSystem/StaticAttributeAgentComponent.cs:
```diff
@@ -0,0 +1,31 @@
+    public class StaticAttributeAgentComponent: AgentComponent
+    {
+        
+        
```
Delete these two newlines on 10 and 11 please

tristanka commented on CSharpSourceCode/AttributeDataSystem/StaticAttributeMissionLogic.cs:
```diff
@@ -0,0 +1,125 @@
+        }
+        
+         public override void OnAgentBuild(Agent agent, Banner banner)
+        {
```
Extra space on line 38

tristanka commented on CSharpSourceCode/AttributeDataSystem/StaticAttributeMissionLogic.cs:
```diff
@@ -0,0 +1,125 @@
+        
+        private List<PartyAttribute> defenderAttributes;
+
+        private List<Agent> _agents;
```
Make the spacing between private fields consistent?

tristanka commented on CSharpSourceCode/AttributeDataSystem/StaticAttributeMissionLogic.cs:
```diff
@@ -0,0 +1,125 @@
+            if (agent.IsMount)
+                return;
+            
+            foreach (var partyAttribute in activeAttributes)
```
The level of nesting that goes on in this loop is a bit suspicious to me. Some of this logic should maybe be broken out into private methods to make this more readable.

tristanka commented on CSharpSourceCode/AttributeDataSystem/StaticAttributeMissionLogic.cs:
```diff
@@ -0,0 +1,125 @@
+                        foreach (var attribute in partyAttribute.RegularTroopAttributes)
+                        {
+                            AddStaticAttributeComponent(agent, attribute, partyAttribute);
+                            break;
```
Why do all of these foreach loops break after the first iteration? If it really does need to break, can't you just get rid of the loop entirely?

tristanka commented on CSharpSourceCode/AttributeDataSystem/StaticAttributeMissionLogic.cs:
```diff
@@ -0,0 +1,125 @@
+                    {
+                        if (!agent.IsHero)
+                        {
+                            if (agent.Character.IsSoldier && !partyAttribute.RegularTroopAttributes.IsEmpty())
```
This `if` condition can just be && in line 74 to reduce nesting

tristanka commented on CSharpSourceCode/AttributeDataSystem/StaticAttributeMissionLogic.cs:
```diff
@@ -0,0 +1,125 @@
+                                break;
+                            }
+                        }
+                        else
```
This `else` contains two `if` blocks with no shared logic before or after them. It'd be cleaner to just have 

```
else if (agent.Character.Name == partyAttribute.Leader.Name)
{
    var leaderAttribute = partyAttribute.LeaderAttribute;
    AddStaticAttributeComponent(agent, leaderAttribute, partyAttribute);
    break;
}
else if (!partyAttribute.CompanionAttributes.IsEmpty())
{
    foreach (var companionAttribute in partyAttribute.CompanionAttributes)
    {
        if (agent.Character.Name.ToString() == companionAttribute.id)
        {
            AddStaticAttributeComponent(agent, companionAttribute, partyAttribute);
            break;
        }
    }
}
```

tristanka commented on CSharpSourceCode/Battle/StatusEffects/StatusEffectMissionLogic.cs:
```diff
@@ -28,6 +32,93 @@ public override void OnAgentCreated(Agent agent)
+            
+            
+        }
+
```
This file has a lot of excessive new lines, please delete. Also, did you delete the commented block on 100-108? The conversation was resolved but the block still shows for me. Unpushed commit?

tristanka replied:
Also, what are the empty OnCreated() and OnRenderingStarted() methods for?

tristanka commented on CSharpSourceCode/Battle/StatusEffects/StatusEffectMissionLogic.cs:
```diff
@@ -28,6 +32,93 @@ public override void OnAgentCreated(Agent agent)
+            {
+                foreach (Agent agent in team.TeamAgents)
+                {
+                    if(agent.Character.IsPlayerCharacter||agent.Character.IsHero)
```
Lots of debug code here, it looks like. Shouldn't this be removed?

tristanka commented on CSharpSourceCode/SubModule.cs:
```diff
@@ -106,6 +107,8 @@ protected override void OnGameStart(Game game, IGameStarter gameStarterObject)
             {
                 CampaignGameStarter starter = gameStarterObject as CampaignGameStarter;
+                PartyAttributeManager partyAttributeManager = PartyAttributeManager.Instance;
+                starter.CampaignBehaviors.Add(partyAttributeManager);
```
PartyAttributeManager is accessible through CampaignBehaviors. It's also a singleton. This seems a bit redundant to me. All other campaign behaviors aren't singletons. It'd make more sense to just construct a PartyAttributeManager here and add it to the CampaignBehaviors that way. No reason to make it any more global than we have to.

Z3rca replied:
you are right I removed the singleton property. This was fair enough. For a certain form in between it was necessary but I left it because I was lazy :D

tristanka commented on CSharpSourceCode/Battle/StatusEffects/StatusEffectMissionLogic.cs:
```diff
@@ -28,6 +32,93 @@ public override void OnAgentCreated(Agent agent)
+            var attacker = Mission.Current.Teams.GetAlliesOf(Mission.Current.Teams.Attacker, true);
+            var defender = Mission.Current.Teams.GetAlliesOf(Mission.Current.Teams.Defender, true);
+            
+            var agents =Mission.Current.Agents;
```
Space between = and Mission?

Z3rca replied:
did that and tidied up a bit the empty spaces

tristanka commented on CSharpSourceCode/AttributeDataSystem/PartyAttributeManager.cs:
```diff
@@ -0,0 +1,437 @@
+
+        private void OnPlayerRecruitedTroop(CharacterObject troop, int amount)
+        {
+       //     TOWCommon.Say("Player Troop Recruited");
```
Commented out debug messages should be removed before merging

Z3rca replied:
Besides some I use for later testing and improvements, I reduced most on unnecessary ones

tristanka commented on CSharpSourceCode/AttributeDataSystem/PartyAttributeManager.cs:
```diff
@@ -0,0 +1,437 @@
+            foreach (var attribute in party.RegularTroopAttributes)
+            {
+                if (attribute.id == troop.ToString());
+                return;
```
Indentation

tristanka commented on CSharpSourceCode/AttributeDataSystem/PartyAttributeManager.cs:
```diff
@@ -0,0 +1,437 @@
+                staticAttribute.MagicEffects = new List<string>();
+                staticAttribute.id = troop.ToString();
+                staticAttribute.number = party.PartyBase.MemberRoster.GetTroopCount(troop);
+                party.RegularTroopAttributes.Add(staticAttribute);
```
All the logic here gets repeated right below it. This is unnecessary. Lines 116 to 135 could instead just be

```
StaticAttribute staticAttribute = new StaticAttribute();

staticAttribute.race = "Human";
staticAttribute.status = troop.Occupation == Occupation.Bandit ? "not so ready" : "battle ready";
staticAttribute.MagicEffects = new List<string>();
staticAttribute.id = troop.ToString();
staticAttribute.number = party.PartyBase.MemberRoster.GetTroopCount(troop);
party.RegularTroopAttributes.Add(staticAttribute);

_currentTroop = null;
```

Also, while looking at that, it appears to me that `staticAttribute.number` is an unclearly named field. What is this for? It doesn't seem like it's used anywhere in the code either other than when it's assigned. Can this just be removed from StaticAttribute?

What does the StaticAttribute.status do? Why are Bandits "not so ready" and everybody else is "battle ready"?

Z3rca replied:
number, race status was removed. For now we are just focusing on Attributes like beeing undead , human, and abilities

Z3rca replied:
the attributes got significantly reduced, and only contain a minimum of information.

Z3rca committed:
Merge remote-tracking branch 'origin/development' into CampaignAttributeRebase

to get the branch up to date

Z3rca committed:
working with Agent Extensions

also added a fireball at start for player if wizard is selected.
rogue parties are undead for testing purposes
added a brightwizard ability

Z3rca committed:
fixing fireball ability

Z3rca committed:
changed constructor for single attributes.

they now need a Party Attribute on which they are assgined. Might cause problems later, but in that case I overload the constructor.

Z3rca committed:
Having a Magic User affects now whole party

if Player or Hero is a mage, the whole party is affected

Z3rca committed:
change for past two commits

Z3rca committed:
Winds Of Magic Based Abilities

Fireball requires now 3 Winds of magic, which can be gained during Campaign

Z3rca committed:
forgot to push new added script item to project

Z3rca committed:
Ingame HUD for Winds of Magic

WARNING : Requires update of XML files in TOW_CoreAssets

Z3rca committed:
Cleaned up unnessary items in the static attributes and party attributes

Z3rca committed:
removed comments, and unnessary lines

Z3rca committed:
removed singleton, which seems to be unnecessary

Z3rca committed:
on empty Winds of Magic, Ability is greyed out

Z3rca committed:
removed unnessary code changes on status effect system

Z3rca committed:
cosmetics

- changed id to PartyBaseID
-removed some outcommented and other debug lines

martkartasev commented on CSharpSourceCode/SubModule.cs:
```diff
@@ -128,9 +129,12 @@ protected override void OnGameStart(Game game, IGameStarter gameStarterObject)
                 starter.CampaignBehaviors.RemoveAllOfType(typeof(BackstoryCampaignBehavior));
                 starter.Models.RemoveAllOfType(typeof(CompanionHiringPriceCalculationModel));
                 starter.AddModel(new TowCompanionHiringPriceCalculationModel());
+                starter.AddModel(new CustomBattleMoralModel.TOWCampaignBattleMoralModel());
```
Small thing. Moral vs Morale. Spelling.

Morale is fighting spirit or willingness to carry through with some action.
A moral is relating to the idea of right or wrong.

Z3rca replied:
done

Z3rca committed:
working on custom battle logic for Winds of Magic

- changed setting of Winds of Magic value

Z3rca committed:
changed assignment for Player Party

also differ between custom battle and campaign logic

Z3rca committed:
changed if cases to switch

martkartasev commented on CSharpSourceCode/Battle/AttributeSystem/CustomBattleMoralModel/TOWCampaignBattleMoralModel.cs:
```diff
@@ -0,0 +1,19 @@
+using TaleWorlds.MountAndBlade;
+using TOW_Core.Battle.Extensions;
+
+namespace TOW_Core.Battle.AttributeSystem.CustomBattleMoralModel
```
Fix typo

Z3rca replied:
done

martkartasev commented on CSharpSourceCode/AttributeDataSystem/StaticAttributeMissionLogic.cs:
```diff
@@ -0,0 +1,217 @@
+            base.AfterAddTeam(team);
+            
+           
+        }
```
Refactor unnecessary lines

Z3rca replied:
done

martkartasev reviewed (commented):
There is a bit of refactoring and name correction that should be done. 

Also I feel as if the Winds of Magic should be shown on the Campaign Map in some manner. Ideally they should appear on the bar in the bottom right corner, but I don't know how much that would be.

I think this does work in general, but I don't know what all of the goals are that we are trying to achieve with it.

Z3rca committed:
changed moral to morale

Z3rca committed:
refractoring

Z3rca committed:
clean up, changed custom battle logic
